### PR TITLE
Improve datagrid performance

### PR DESCRIPTION
**Pull request by martinRenou**
cc. @ibdafna 

This PR improves the performance of the grid by reworking the drawing of merged cells logic.

- Before (resizing/scrolling a 1M cells grid with merged cells freezes):

https://user-images.githubusercontent.com/21197331/188858914-860b06bf-4005-4439-9814-93150ee4e5cc.mp4

- After (resizing/scrolling a 1M cells grid with merged cells works smoothly):

https://user-images.githubusercontent.com/21197331/188859052-f09e20a2-dac7-46a4-b985-4b9cbaf0238e.mp4

Note for the tester: You will see a big difference in performance if you zoom out a lot and show a lot of cells.

### CODE REVIEW

File: packages/datagrid/src/datagrid.ts [hunk 1]
```diff
@@ -2146,7 +2144,7 @@ export class DataGrid extends Widget {
     }
 
     // Test whether there is content to blit.
-    let needBlit = curH > 0 && curH > 0 && width > 0 && height > 0;
+    let needBlit = curW > 0 && curH > 0 && width > 0 && height > 0;
```
Comment (afshin) [inline packages/datagrid/src/datagrid.ts:2147]: # `🕵️ 🔍 🐜`

**Comment by martinRenou** [comment]
This is the diff for the API:
```diff
diff --git a/review/api/datagrid.api.md b/review/api/datagrid.api.md
index 9197a504..ebd1c1c3 100644
--- a/review/api/datagrid.api.md
+++ b/review/api/datagrid.api.md
@@ -131,16 +131,11 @@ export interface CellGroup {
 export namespace CellGroup {
     export function areCellGroupsIntersecting(group1: CellGroup, group2: CellGroup): boolean;
     export function areCellGroupsIntersectingAtAxis(group1: CellGroup, group2: CellGroup, axis: 'row' | 'column'): boolean;
-    // (undocumented)
-    export function areCellsMerged(dataModel: DataModel, rgn: DataModel.CellRegion, cell1: number[], cell2: number[]): boolean;
-    export function calculateMergeOffsets(dataModel: DataModel, regions: DataModel.CellRegion[], axis: 'row' | 'column', sectionList: SectionList, index: number): [number, number, CellGroup];
     export function getCellGroupsAtColumn(dataModel: DataModel, rgn: DataModel.CellRegion, column: number): CellGroup[];
     export function getCellGroupsAtRegion(dataModel: DataModel, rgn: DataModel.CellRegion): CellGroup[];
     export function getCellGroupsAtRow(dataModel: DataModel, rgn: DataModel.CellRegion, row: number): CellGroup[];
     export function getGroup(dataModel: DataModel, rgn: DataModel.CellRegion, row: number, column: number): CellGroup | null;
     export function getGroupIndex(dataModel: DataModel, rgn: DataModel.CellRegion, row: number, column: number): number;
-    export function isCellGroupAbove(group1: CellGroup, group2: CellGroup): boolean;
-    export function isCellGroupBelow(group1: CellGroup, group2: CellGroup): boolean;
     export function joinCellGroups(groups: CellGroup[]): CellGroup;
     export function joinCellGroupsIntersectingAtAxis(dataModel: DataModel, regions: DataModel.CellRegion[], axis: 'row' | 'column', group: CellGroup): CellGroup;
     export function joinCellGroupWithMergedCellGroups(dataModel: DataModel, group: CellGroup, region: DataModel.CellRegion): CellGroup;
```

@afshin @blink1073 do you think that's a problem? I can add those back and add a DeprecationWarning if that helps the PR get merged.

**Comment by martinRenou** [comment]
A bi-product of this PR is that it fixes a rendering issue when shrinking cells that are at the same level as a merged cell:

- Before:

https://user-images.githubusercontent.com/21197331/188902742-00a6ec2b-9463-4c7d-aac0-66d85b983e9f.mp4

- After:

https://user-images.githubusercontent.com/21197331/188902778-62692db1-eecf-4b3f-9bdd-0c6e6d1eb39f.mp4

**Comment by blink1073** [comment]
Removing them seems reasonable to me since datagrid is still pre-1.0.

**Comment by martinRenou** [comment]
> Removing them seems reasonable to me since datagrid is still pre-1.0.

Thanks for your answer! I updated the API file

**Review by afshin — APPROVED** [review]
Nice work tracking down the performance issue and resolving it, @martinRenou!

I think we can merge this and iterate because:
- This is a pre-1.0 package (although, perhaps we should consider jumping straight to 2.0 to be in line with the rest of the library)
- `@lumino/datagrid` is downstream of the other packages

But we should make sure that before the final version we address:
- refactoring with a `type PaintRegion`
- implementing and removing the `TODO` messages
- potentially de-duplicating the slightly different but very similar loops

**Comment by martinRenou** [comment]
> But we should make sure that before the final version we address:
> - refactoring with a type PaintRegion
> - implementing and removing the TODO messages
> - potentially de-duplicating the slightly different but very similar loops

Definitely 👍🏽 

Maybe @ibdafna would like to give this a try before we merge? As he implemented the merge cells logic and knows what to test etc to make sure everything works nicely.

**Comment by ibdafna** [comment]
Thank you for this @martinRenou, I will test tomorrow evening and report
back. I have no doubt this is working incredibly well! 🙏

**Comment by ibdafna** [comment]
https://user-images.githubusercontent.com/24281433/189535645-0f32fb6c-8120-49d3-b212-66734125ed8a.mp4

@martinRenou I see some issues with border rendering when scrolling x/y axes (video attached)

**Comment by martinRenou** [comment]
Thanks for trying! I haven't seen those. I will try to replicate.

**Comment by martinRenou** [comment]
@ibdafna the latest commit fc7acd4 is fixing the issue you mention above :) 

https://user-images.githubusercontent.com/21197331/191047999-7368b342-a389-4ea9-9bad-213171787a4e.mp4

**Comment by martinRenou** [comment]
> Looking at my screencast I now see an issue with the column header C 0, 4 which is not visible (I can reproduce the issue and working on it)

This is fixed by 32952d7. The issue was that I was taking into account the x-scroll level for the column header and the y-scroll level for the row header while this should not be the case (headers are sticky, not following the scroll).

**Comment by martinRenou** [comment]
@afshin @ibdafna this is ready for another review :)

**Comment by ibdafna** [comment]
LGTM! Let's ship it!! ♥💘💖💗💓💙💚💛❤🖤🧡💜

**Comment by martinRenou** [comment]
Thanks!! I actually found another issue when slowly scrolling. 830e094 seems to fix that issue. 

> potentially de-duplicating the slightly different but very similar loops

The improvement suggested by @afshin gets more urgent to fix as it's now 6 loops that are quite similar, but I can do it in a separate PR if that's fine.

**Comment by ibdafna** [comment]
I'd love to see this released this week 😻

**Review by afshin — APPROVED** [review]
Thank you! 🚀

**Pull request merged**